File: c_programming/array/n48_find_first_and_last_position_of_element_in_sorted_array_34.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include "utils.h"

static int32_t find_first(const int32_t *array, size_t sz, int32_t target, int32_t *out)
{
    int32_t ret = 0;
    int32_t left = 0, right = sz - 1;
    int32_t mid = left + (right - left) / 2;
    int32_t i = 0;
    bool is_left = false;
    bool is_right = false;

    UTILS_CHECK_PTR(array);
    UTILS_CHECK_PTR(out);
    out[0] = -1;
    out[1] = -1;

    UTILS_CHECK_LEN(sz);

    while (left <= right) {
        if (array[mid] < target) {
            left = mid;
        } else if (array[mid] > target) {
            right = mid;
        } else {
            break;
        }
        mid = left + (right - left) / 2;
    }

    if (mid == 0 || mid == sz - 1 ) {
        goto finish;
    }

    left = right = mid;
    while (is_left == false || is_right == false) {
        if (array[mid] == array[left]) {
            left --;
        } else {
            out[0] = left + 1;
            is_left = true;
        }

        if (array[mid] == array[right]) {
            right ++;
        } else {
            out[1] = right - 1;
            is_right = true;
        }
    }

finish:
    return ret;
}

int32_t main(void)
{
    int32_t ret = 0;
    int32_t array[] = {5,7,7,7,7,7,7,7,7,7,8,8,8,8,8,8,8,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,10};
    int32_t out[2] = {0};

    ret = find_first((const int32_t *)array, ARRAY_SIZE(array), 8, out);
    UTILS_CHECK_RET(ret);

    LOG("the output is [%d, %d]\n", out[0], out[1]);

finish:
    return ret;
}

File: rust_programming/array/src/n48_find_first_and_last_position_of_element_in_sorted_array_34.rs

fn find_first(array:&Vec<i32>, target:i32) -> Vec<i32>
{
    let mut ret_vec:Vec<i32> = vec![-1, -1];
    let mut left:usize = 0;
    let mut right:usize = array.len() - 1;
    let mut mid:usize = left + (right - left) / 2;

    while left <= right {
        if array[mid] < target {
            left = mid;
        } else if array[mid] > target {
            right = mid;
        } else {
            break;
        }
        mid = left + (right - left) / 2;
    }
    
    if (mid == 0) || (mid == array.len() - 1) {
        return ret_vec;
    }

    right = mid;
    left = mid;
    let mut is_left:bool = false;
    let mut is_right:bool = false;
    while (is_left == false) || (is_right == false) {
        if array[mid] == array[left] {
            left -= 1;
        } else {
            ret_vec[0] = (left + 1) as i32;
            is_left = true;
        }

        if array[mid] == array[right] {
            right += 1;
        }  else {
            ret_vec[1] = (right - 1) as i32;
            is_right = true;
        }
    }

    return ret_vec;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    pub fn test_shortest_distance() {
        let array:Vec<i32> = vec![5,7,7,7,7,7,7,7,7,7,8,8,8,8,8,8,8,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,10];
        let target:i32 = 8;
        let ret = find_first(&array, target);

        assert_eq!(ret, vec![10, 16]);
    }
}